providers/dns/autodns/autodns.go

```diff
@@ -0,0 +1,106 @@
+package autodns
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/go-acme/lego/v3/challenge/dns01"
+	"github.com/go-acme/lego/v3/platform/config/env"
+)
+
+const (
+	envAPIUser     = `AUTODNS_API_USER`
+	envAPIPassword = `AUTODNS_API_PASSWORD`
+	envAPIEndpoint = `AUTODNS_ENDPOINT`
+
+	defaultEndpoint = `https://api.autodns.com/v1/`
+	demoEndpoint    = `https://api.demo.autodns.com/v1/`
+
+	defaultEndpointContext int = 4
+)
+
+type Config struct {
+	Endpoint   *url.URL
+	Username   string `json:"username"`
+	Password   string `json:"password"`
+	Context    int    `json:"-"`
+	HTTPClient *http.Client
+}
+
+func NewDefaultConfig() *Config {
+	endpoint, _ := url.Parse(defaultEndpoint)
+
+	return &Config{
+		Endpoint:   endpoint,
+		Context:    defaultEndpointContext,
+		HTTPClient: &http.Client{},
+	}
+}
+
+type DNSProvider struct {
+	config *Config
+	//zoneNameservers map[string]string
+	//currentRecords  []*ResourceRecord
+}
+
+func NewDNSProvider() (*DNSProvider, error) {
+	values, err := env.Get(envAPIUser, envAPIPassword)
+	if err != nil {
+		return nil, fmt.Errorf("autodns: %v", err)
+	}
+
+	rawEndpoint := env.GetOrDefaultString(envAPIEndpoint, defaultEndpoint)
+	endpoint, err := url.Parse(rawEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("autodns: %v", err)
+	}
+
+	config := NewDefaultConfig()
+	config.Username = values[envAPIUser]
+	config.Password = values[envAPIPassword]
+	config.Endpoint = endpoint
+
+	provider := &DNSProvider{config: config}
+
+	// Because autodns needs the nameservers for each request, we query them all here and put them
+	// in our state to avoid making a lot of requests later.
+	// FIXME: This should become obsolete once I figure out how the _stream endpoint works.
```

Comment:
did you ask AutoDNS developers?

Reply:
Yes, I sent them a mail.

Reply:
I used it like in their documentation, but it did not work, so we'll just wait what they have to say.
The challenge works though if you check out 1b8f6763d55523bb27e8329055b39b708dcc6cc2